### DISCUSSION TITLE
Extract FSM Does Not Work

### DISCUSSION THREAD

**Discussion by myzinsky** [YosysHQ/yosys#2951] [Q&A]
Following the example here:

https://stackoverflow.com/questions/32645965/fsm-export-using-yosys/65638821#65638821

``` bash
yosys -p 'read_verilog fsm.v; proc; opt; fsm -nomap; fsm_export -o test.kiss2'
```

does not export any kiss2 file :-(

Is that a bug? Or has the usage changed?

**Comment by Ravenslofty** [comment]
This could be #2402; if there are no FSMs detected, then `fsm_export` does nothing.

**Comment by MJChku** [comment]
Another answer is don't run opt;  Just run proc; fsm ...

**Comment by nakengelhardt** [comment]
`fsm_detect` expects the following passes to be run beforehand: `hierarchy -top <top>; proc; bmuxmap; opt -nodffe -nosdff`

This is because `fsm_detect` works on a limited subset of FF cells, and also doesn't handle `$bmux` cells. These passes give the FSM detection heuristic the shape of design it expects.
